// pgsql-parser/main/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Deparser", {
  enumerable: true,
  get: function get() {
    return _pgsqlDeparser.Deparser;
  }
});
Object.defineProperty(exports, "deparse", {
  enumerable: true,
  get: function get() {
    return _pgsqlDeparser.deparse;
  }
});
Object.defineProperty(exports, "enums", {
  enumerable: true,
  get: function get() {
    return _pgsqlEnums.enums;
  }
});
Object.defineProperty(exports, "parseFunction", {
  enumerable: true,
  get: function get() {
    return _libpgQuery.parsePlPgSQLSync;
  }
});
exports.parse = void 0;

var _pgsqlDeparser = require("pgsql-deparser");

var _pgsqlEnums = require("pgsql-enums");

var _libpgQuery = require("libpg-query");

var parse = function parse(sql) {
  if (!sql) throw new Error('no SQL provided to parser');
  var result = (0, _libpgQuery.parseQuerySync)(sql);
  return result.stmts.map(function (_ref) {
    var stmt = _ref.stmt,
        stmt_len = _ref.stmt_len;
    return {
      RawStmt: {
        stmt: stmt,
        stmt_len: stmt_len
      }
    };
  });
};

exports.parse = parse;
